feat(api): nvim_echo can emit Progress messages/events

shadmansaleh opened the pull request:
Working toward #32537

based on a proposal from https://github.com/neovim/neovim/issues/32537#issuecomment-2672600804


## Problem:
Nvim does not have a core concept for indicating "progress" of
long-running tasks. The LspProgress event is specific to LSP.

## Solution:
- `nvim_echo` can emit `kind="progress"` messages.
  - Emits a `Progress` event.
  - Includes new fields (id, status, percent) in the `msg_show` ui-event.
  - The UI is expected to overwrite any message having the same id.
- Messages have a globally unique ID.
  - `nvim_echo` returns the message ID.
- `nvim_echo(… {id=…})` updates existing messages.

## Example:

```lua
local grp = vim.api.nvim_create_augroup("Msg", {clear = true})
vim.api.nvim_create_autocmd('Progress', {
  pattern={"term"},
  group = grp,
  callback = function(ev)
    print(string.format('event fired: %s', vim.inspect(ev))..'\n')
  end
})

-- require('vim._extui').enable({enable=true, msg={target='msg', timeout=1000}})
vim.api.nvim_echo({{'searching'}}, true, {kind='progress',  percent=80, status='running', title="terminal(ripgrep)"})
local id = vim.api.nvim_echo({{'searching'}}, true, {kind='progress', status='running', percent=10, title="terminal(ripgrep)"})
vim.api.nvim_echo({}, true, {id = id, kind='progress', percent=20, status = 'running', title='find tests'})
vim.api.nvim_echo({}, true, {id = id, kind='progress', status='running', percent=70})
vim.api.nvim_echo({{'complete'}}, true, {id = id, kind='progress', status='success', percent=100, title="find tests"})
```

## Todo:
- [x] ~add Message event for messages outside of history too~
       A general message event is really complicated without a proper refactor to message/echo-area. Particularly, non-multiline message/vim stuff touches this section in too many ways at too many places. Considering it out of scope for now.
       Introducing a `Progress` event for progress messages instead of a general message event.
- [x] integrate with `msg_show` ui event
- [x] testing
- [x] docs

## Followups:
- Integrate with 'statusline' by listening to the Progress autocmd event.
- Integrate progress ui-event with `vim._extui`.

shadmansaleh committed:
wip

shadmansaleh committed:
wip2

shadmansaleh commented on src/nvim/message_defs.h:
```diff
@@ -10,14 +11,19 @@ typedef struct {
 typedef kvec_t(HlMessageChunk) HlMessage;
+typedef uint64_t MsgID;
 
+#define MSG_KIND_PROGRESS "progress"
```
Where are UI protocols kept actually? What makes a sting valid message-kind?

luukvbaal replied:
Not sure what you're asking sorry. Until now message kind strings are just set directly at the callsite, not stored/defined anywhere. One exception is the verbose kind to compare to `msg_ext_kind` directly rather than needing `strcmp()`
https://github.com/neovim/neovim/blob/435f03ee106a548c2e5fd6c10005f5ae2a29a5a4/src/nvim/message.c#L3325-L3334

shadmansaleh replied:
I noticed that. But I was wondering there should be at least some validation layer. Otherwise, kind can be set to anything. I guess that's the case here. The verbose case is different it lt's controlled by verbose_enter() and verbose_exit() that's why it can work with pointer comparison. It doesn't make sense for progress to be something like that.

shadmansaleh committed:
wip-fix crash

shadmansaleh committed:
fix ci build

shadmansaleh committed:
fix test

shadmansaleh committed:
fix lint

shadmansaleh committed:
add title

shadmansaleh committed:
allow all messages to be updated

assign default behaviors for status and title
when new one is not given past one is kept

shadmansaleh committed:
fixups

shadmansaleh commented on src/nvim/api/vim.c:
```diff
@@ -778,7 +778,21 @@ void nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history, Dict(
+  String status = opts->status;
+  if ((kind != NULL && strcmp(kind, "progress") == 0)
+      && ((status.data == NULL)
+          || (strcmp(status.data, "success") != 0
```
What would be good candidate options for status? also if status is failed should we mark it as error message?

shadmansaleh commented on src/nvim/message.c:
```diff
@@ -1034,16 +1053,50 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
+  MessageHistoryEntry *entry = msg_find_by_id(msg_id);
+  if (entry) {
+    progress_message_found = true;
+    // detach the node if found
```
When a message gets updated should we drop the message to end of the history? or leave it where it is?

luukvbaal replied:
Why bother manipulating the message history at all? Can we not just add a new entry for each message (if `history == true`, which it arguably shouldn't be for "progress" messages at least).

Being able to replace a message is about replacing it in the UI right? How the history looks like seems insignificant, and may even be unwanted?

luukvbaal replied:
Yes I'm also not sure this needs to leak into the C side in the extent it is currently proposed. I would think just assigning a message ID, which is then passed along the the `msg_show` event, is sufficient. A UI can then replace the message text in case the message is still visible somewhere.

What is the purpose of the title, status and percentage fields? Can the caller not represent all of those by modifying the message chunks?

shadmansaleh replied:
> Can the caller not represent all of those by modifying the message chunks?

Technically yes. With kind being able to be anything client can even just drop e json blob in the msg and ui can handle it. But since we are adding explicit support for progress message internally I think it makes sense to keep them as separate fields. When we through autocmd events we can provide them separately then receiver can deal with it how it wants.

shadmansaleh committed:
fix tests

shadmansaleh committed:
restrict msg title, status, progress to progress messages only

shadmansaleh commented:
I have a few questions

- in the proposal does the  ui-events (msg_show) or autocmd-events (like a new Message event) ?
- what's our policy on ui-events is adding more arguments a breaking change to the protocol?
- actually I'm also confused on how does ui-events get sents there is a function that constracts snd sends the event but I could't find where it gets called in nvim_echos code path. Unless msg_start() and msg_end() are triggering it. I guess I need to explore more of the ext-ui codes. I'm comparativly new to that section of the codebase.
- should we make progress messages always to be recorded in history? Since we only save messages in message history. I'd assume users can send partial updates in future messages. If we want the events we emit to have full info then we need to store the messages
- when a message update comes should we drop the message node to end of history. Or modify in where it is.
- should we let progress percent to be decreased? Or allow it only to increase? In case of vscode they provide a increment option only.
- also in your proposal what does status actually represent?

cc @justinmk

justinmk commented:
1. editor event (autocmd)
1. adding new args to an event is non-breaking (`:help api-contract`)
1. UI events are asynchronous and not related to any particular function. `nvim_echo` happens to affect the UI (it prints messages) so it generates UI events.

> should we make progress messages always to be recorded in history?

I was thinking about this today and starting to doubt some portion of my proposal. Because progress messages are long-lived, and when they are updated they generate messages, so it doesn't really make sense to "update" an existing message. That would make the message system logic very strange.

> when a message update comes should we drop the message node to end of history. Or modify in where it is.

yeah, that's the part I am rethinking. Pushing it to the end would increment the id...? But that makes no sense. And if it doesn't increment the id, then the message system has to reason about the order of messages.

> should we let progress percent to be decreased? Or allow it only to increase? In case of vscode they provide a increment option only.

Is it a reversible decision? If so, we can choose the simple way.

> also in your proposal what does status actually represent?

`status='done|fail|cancel|waiting|...'`

shadmansaleh commented:
> UI events are asynchronous and not related to any particular function. nvim_echo happens to affect the UI (it prints messages) so it generates UI events.

The `msg_show` UI event for should emit message_id, status, title, percent stuff too. I was trying to figure out how to do it.

> Pushing it to the end would increment the id...? But that makes no sense. And if it doesn't increment the id, then the message system has to reason about the order of messages.

As the current implementation in this pr stands it doesn't increment the id but does move the message to the end. We can do this because message-history is stored as linked-list. Linked-list has it's own ordering apart from the id we are assigning a message.

I'm kind of confused if message history is the right place to implement this. We currently only give message-id to messages when messages are stored in history. `nvim_echo` can create temporary messages that doesn't get stored in message history. Then there's also ext-ui.

> Is it a reversible decision? If so, we can choose the simple way.

I'm not sure what you meant here.

luukvbaal commented:
Like I said, simply passing the ID to the msg_show event (emitted [here](https://github.com/neovim/neovim/blob/3e7f5d95aa25943e26d88a75fc55785229c36e34/src/nvim/message.c#L3183), modified [here](https://github.com/neovim/neovim/blob/3e7f5d95aa25943e26d88a75fc55785229c36e34/src/nvim/api/ui_events.in.h#L165)) would be sufficient for a UI to replace the message. The UI itself would track whether it is still visible, doesn't need to be attached to the message history in `message.c`.

shadmansaleh commented:
> simply passing the ID to the msg_show event (emitted here, modified here) would be sufficient for a UI to replace the message

Not exactly. That takes care of `ext-ui` clients but the tui won't handle things just with that. Tui would just be printing the message as is.

> (adding status, title, percent, to the UI event still seems excessive to me).
I'm guessing you want the progress-msg emiter to handle how they want to format things or how they want to display things as part of the message text. That makes the entire thing way too openended. Creating seperation like this allows us to standardize things. Also create separation. For example, title, status and percent are special info for progress messages. Yeah you can just format the message text like "title: msg percent%" but there's no enforcement for that. And it can't be reliably decomposed due to that. If we keep them as separate things UI can decide to deal with it differently. Also, since we are going to run special autocmd on progress updates plugins can listen for those too. The `status` with events can also act as cancelation trigger.

> The UI itself would track whether it is still visible, doesn't need to be attached to the message history in message.c.

Message history integration is mainly to keep track of the messages as internal states, to handle tui and to basically allow partial updates. Like you don't change anything in the messages/title/status just bump progress percent to 10 to 20% l

przepompownia commented:
Will the `replace_last` parameter become redundant after providing message ids?

shadmansaleh commented:
> It will become a UI client.

Since it's not yet I think it makes sense to support the current tui. It's not a ton of code anyway. If it becomes later down the line it can just be removed then.

> maybe we should consider adding a data field rather than 4 new arguments.

By single field data do you mean a subobject/map or just one string field? If the later then same issue with separation is still there. Also, in the 1st casse it's already in an object why create an extra level of nesting? Internally we can handle it as a seperate object but externally in api/events it kind of stays the same

shadmansaleh committed:
wip aucmd

shadmansaleh committed:
refactor + add Message autocmd-event

shadmansaleh committed:
additional validation

shadmansaleh committed:
fix leak?

shadmansaleh committed:
fixup

shadmansaleh committed:
emit percent status and title only on progress messages

shadmansaleh committed:
send progress event for progress messages instead of general message events

shadmansaleh committed:
fix lint

shadmansaleh committed:
special formatting for displaying progress messages

shadmansaleh committed:
refactor + test fix?

shadmansaleh committed:
fixup

shadmansaleh committed:
fix verbose test fail again :]

shadmansaleh committed:
docs: update docs

shadmansaleh committed:
fix lint

shadmansaleh committed:
emit progress info in msg_show ui event

shadmansaleh committed:
update docs

shadmansaleh committed:
simplify logic

shadmansaleh committed:
don't emit special progress msg formatting to ext-ui

shadmansaleh committed:
rename id field to msg_id in progress event since there is already an id field for autocmd

shadmansaleh committed:
fix correct msg-id not being send to ext-ui

shadmansaleh committed:
add tests

shadmansaleh committed:
remove redundent imports

shadmansaleh committed:
make status optional, validate percent is 0-100, let error msg as progress

shadmansaleh committed:
fix lint

shadmansaleh commented:
@justinmk, I think it's in good shape now. A review here will be nice. Also, take a look at the opening message of the pr. It explains how things work.

shadmansaleh commented:
> @justinmk, I think it's in good shape now. A review here will be nice. Also, take a look at the opening message of the pr. It explains how things work.

Ping @justinmk

justinmk commented on runtime/doc/api-ui-events.txt:
```diff
@@ -877,6 +878,24 @@ must handle.
+	    to 1. 0 indicates msg_id was not assigned. When message of same id
+	    appears it should replace the older message.
+
+	extra_info
```
Let's call this `progress`. All the fields are for that purpose. If we need a free-form field (`data`) we can always add it later.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -787,6 +787,32 @@ ModeChanged			After changing the mode. The pattern is
+Progress						*Progress*
+				After a progress-msg is message is created via
+				`nvim_echo`. The pattern is matched against
+				title of the message. The data item in autocmd
```
matching against the title is unusual but makes sense I think

shadmansaleh replied:
I was thinking of matching against something like ID:TITLE . That might be more useful.

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -787,6 +787,32 @@ ModeChanged			After changing the mode. The pattern is
+Progress						*Progress*
+				After a progress-msg is message is created via
+				`nvim_echo`. The pattern is matched against
+				title of the message. The data item in autocmd
+				callback contains following
```
we have a tag for that
```suggestion
				title of the message. The |event-data| contains:
```

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -787,6 +787,32 @@ ModeChanged			After changing the mode. The pattern is
+				`nvim_echo`. The pattern is matched against
+				title of the message. The data item in autocmd
+				callback contains following
+					msg_id: id of the message
+					message: text of the message
```
use either "msg" or "message", not both ways. but maybe `text` could work.
```suggestion
					msg_id: id of the message
					text: message text
```

shadmansaleh replied:
Should we use `content` instead? ui-msg-show event uses `content` parameter for this

justinmk replied:
ui event param names because they are not permanent (they are ordered args, not named). they are not a great source of best practices.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -760,12 +760,29 @@ void nvim_set_vvar(String name, Object value, Error *err)
+///          - percent: How much progress is done on the progress
+///            message
+/// @return The message-id of the message.
+///         valid message-id is always greater or equal to 1
+///         -1 means nvim_echo didn't show a message
+///          0 means nvim_echo didn't allocate a message id for the message. happens
+///            for temp messages not stored in message history.
```
does using bullet points improve the generated docs? we don't need to mention `nvim_echo` in so many places.

```suggestion
///         valid message-id is always greater or equal to 1
///         - -1 no message was shown
///         -  0 temp message (not stored in history)
```

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -780,7 +797,40 @@ void nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history, Dict(
-  msg_multihl(hl_msg, kind, history, opts->err);
+  bool is_kind_progress = kind != NULL && strcmp(kind, "progress") == 0;
+
+  if (!is_kind_progress && (opts->status.size != 0 || opts->title.size != 0 || opts->percent > 0)) {
+    api_set_error(err, kErrorTypeValidation,
+                  "title, status and percents fields can only be used with progress messages");
```
use one of the VALIDATE macros. same for most other `api_set_error` cases below, except in rare cases.

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -780,7 +797,40 @@ void nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history, Dict(
+
+  if (is_kind_progress
+      && ((opts->status.data != NULL)
+          && (strcmp(opts->status.data, "success") != 0
```
`strequal` avoids some verbosity, and it also allows NULL

justinmk commented on src/nvim/message.c:
```diff
@@ -1034,16 +1133,62 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
-  // Allocate an entry and add the message at the end of the history.
-  MessageHistoryEntry *entry = xmalloc(sizeof(MessageHistoryEntry));
-  entry->msg = msg;
+  bool is_kind_progress = msg_ext_kind != NULL && strcmp(msg_ext_kind, MSG_KIND_PROGRESS) == 0;
```
use strequal() wherever appropriate, especially when you are checking "not NULL"

```suggestion
  bool is_progress = strequal(msg_ext_kind, MSG_KIND_PROGRESS);
```

justinmk commented on src/nvim/message.c:
```diff
@@ -160,6 +164,16 @@ static bool msg_ext_history = false;  ///< message was added to history
+static MsgID msg_id_next = 1;           ///< message id to be allocated to next message
+
+// code inside doesn't emit ext-message
+#define NO_EXT_MSG(code) \
+  do { \
+    msg_ext_skip = true; \
+    code; \
+    msg_ext_skip = false; \
+  } while (0)
```
this looks very sus. why is this necessary for progress messages but not other messages?

if we drop `draw_progress_message` and only support "extui", can we get rid of this? 

even without "extui", couldn't we show progress message by handling a `Progress` event in Lua ?

shadmansaleh replied:
This is mainly because of this

> - Progress message get's displayed in TUI/cmdline in the following format
`{title}: {msg}..{percent}%`

In tui I'm showing the msg in special format to include title and percent. In case of ext-ui events I'm sending those segments seperately in msg_show event. Basically whatever we write on tui msg_area it get's forwarded to ext-ui as msg. That's is handled quite more lower-level in code. Passing around a flag as arg seemed wayy more messy then this to me.

This basically just adds a file-local global flag that enables and disables that forwarding the macro just wrap the code with toggle for the flag.

Basically if we get rid of it we'll only show the msg without title and percent. Extui will still get all the info and it can show as they like

justinmk commented on src/nvim/message.c:
```diff
@@ -1060,10 +1205,29 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
   }
 
-  msg_hist_len += !temp;
+  msg_hist_len += !temp && !old_message_found;
```
lots of new flags and special conditions are being added in this PR. please find a way to reduce them.

shadmansaleh commented on src/nvim/api/vim.c:
```diff
@@ -780,7 +797,40 @@ void nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history, Dict(
   }
 
-  msg_multihl(hl_msg, kind, history, opts->err);
+  bool is_kind_progress = kind != NULL && strcmp(kind, "progress") == 0;
```
Wasn't aware of strequal. Will update with that

shadmansaleh committed:
address review-1

shadmansaleh committed:
drop special msg pattern in tui-messages

shadmansaleh committed:
drop partial updates

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -181,6 +182,9 @@ EVENTS
 
 • |CmdlineLeavePre| triggered before preparing to leave the command line.
 • New `append` paremeter for |ui-messages| `msg_show` event.
+• New `msg-id` and `extra-info` paremeter for |ui-messages| `msg_show` event.
```
This doesn't match api-ui-events.txt

shadmansaleh replied:
Names changed a bit from review. Forgot to update it here. Will do.

zeertzjq commented on src/nvim/auevents.lua:
```diff
@@ -89,6 +89,7 @@ return {
     QuitPre = false, -- before :quit
     PackChangedPre = false, -- before trying to change state of `vim.pack` plugin
     PackChanged = false, -- after changing state of `vim.pack` plugin
+    Progress = false, -- after showing/updating  a progress message
```
```suggestion
    Progress = false, -- after showing/updating a progress message
```

zeertzjq commented:
The documentation seems a bit confusing. The term "progress message" appears multiple times, but it is never defined.

shadmansaleh committed:
update docs

shadmansaleh committed:
remove redundant changes

echasnovski commented on runtime/doc/api-ui-events.txt:
```diff
@@ -909,4 +928,29 @@ must handle.
+    • Updating or creating a progress message also triggers the |Progress| autocommand.
+
+Example: >
+    local id = vim.api.nvim_echo({{'searching...'}}, true,
+          {kind='progress', status='running', percent=10, title="term"})
+    vim.api.nvim_echo({{'searching'}}, true,
+          {id=id, kind='progress', status='running', percent=50, title="term"})
```
> Asking because for LSP there is a `lsp.ProgressToken integer|string` which is used to identify progress messages. The token is provided from the servers. If `nvim_echo` forces us to use the `msg-id` it creates then it means the lsp layer would have to maintain a progressToken->msg-id mapping.

Externally created message id are not guaranteed to be unique, i.e. they might conflict with other externally or future automatically created message ids. For careful usages should not be a problem, but still. So at least it might require some id validation or something.

For example, `nvim_buf_set_extmark()` allows setting manual numeric `id`, but it adjusts how new ids are computed based on that.

shadmansaleh replied:
I was mainly thinking about conflicts myself. For me, the ID was sort of like the namespace concept we have in other APIs. If I remember right, the VSCodes system allows you to pass in UUIDs, not just int as string, lsp, and dap specs, probably inherited the string|integer from that. Here we are also marking all messages saved in history with an ID, not just progress messages. This is primarily providing the low-level api for isn't making it simpler better? Any user of the api can do the translation in a higher level if necessary. Also, do we do multiple types of args in api? like int/str? I guess it can be done. But it's just bringing in complexity to lower-level.

mfussenegger replied:
> . This is primarily providing the low-level api for isn't making it simpler better? Any user of the api can do the translation in a higher level if necessary

Would the higher level API be in core as well?
There will be N providers - each plugin that wants to report progress in some way, but only a single `nvim_echo`. Imho moving complexity from the N locations to the single location is worthwhile. 


> Also, about the userdata, do you know what's usually kept there?

It would be mostly about being able to provide additional data that users might handle in their `Progress` handlers.

LSP uses these structures:

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workDoneProgressBegin

DAP these:

https://microsoft.github.io/debug-adapter-protocol/specification#Events_ProgressStart


I suspect that one goal of this is to adapt https://github.com/neovim/neovim/blob/30dae87de4c6d6c8bda9657e22e187634cfa12a8/runtime/lua/vim/lsp/handlers.lua?plain=1#L56-L87 to go through nvim_echo?

shadmansaleh committed:
ensure status is given

shadmansaleh committed:
fix test

shadmansaleh committed:
allow arbitary data in nvim_echo through extra_info

shadmansaleh committed:
fix typo in docs

shadmansaleh committed:
turn vim.pack use of nvim_echo to non progress for now

shadmansaleh committed:
fix lint

shadmansaleh committed:
fix leaks

shadmansaleh commented:
ping @justinmk

zeertzjq commented on runtime/doc/api-ui-events.txt:
```diff
@@ -846,6 +846,7 @@ must handle.
 		"lua_error"	Error in |:lua| code
 		"lua_print"	|print()| from |:lua| code
 		"rpc_error"	Error response from |rpcrequest()|
+		"progress"	Progress message emitted by `nvim_echo`
```
```suggestion
		"progress"	Progress message emitted by |nvim_echo()|
```

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -189,6 +190,9 @@ EVENTS
 • |CmdlineLeavePre| triggered before preparing to leave the command line.
 • New `append` paremeter for |ui-messages| `msg_show` event.
+• New `msg_id` and `progress` paremeter for |ui-messages| `msg_show` event.
+• Creating or updating a progress message with `nvim_echo` triggers a |Progress| event.
+
```
```suggestion
• Creating or updating a progress message with |nvim_echo()| triggers a |Progress| event.
```

shadmansaleh committed:
address review (doc-fixups)

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -189,6 +190,9 @@ EVENTS
 • New `append` paremeter for |ui-messages| `msg_show` event.
+• New `msg_id` and `progress` paremeter for |ui-messages| `msg_show` event.
+• Creating or updating a progress message with |nvim_echo| triggers a |Progress| event.
+
```
```suggestion
```

shadmansaleh committed:
doc fixup 2

justinmk commented on runtime/doc/api-ui-events.txt:
```diff
@@ -846,6 +846,7 @@ must handle.
 		"lua_error"	Error in |:lua| code
 		"lua_print"	|print()| from |:lua| code
 		"rpc_error"	Error response from |rpcrequest()|
+		"progress"	Progress message emitted by |nvim_echo()|
```
this list is almost sorted, can you move this up 1 line so that the list stays almost-sorted :)

justinmk commented on runtime/doc/api-ui-events.txt:
```diff
@@ -881,6 +882,24 @@ must handle.
+	    appears it should replace the older message.
+
+	progress
+	    Placeholder for special items for special messages. Currently, used
+	    for progress-messages. Can contain following for progress-message
```
unnecessary preamble
```suggestion
	    Progress-message properties:
```

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -758,12 +758,30 @@ void nvim_set_vvar(String name, Object value, Error *err)
+///          - percent: How much progress is done on the progress
+///            message
+///          - extra_info: dictionary containing additional information
+/// @return The msg-id of the message.
```
```suggestion
/// @return Message id.
```

justinmk commented on src/nvim/message_defs.h:
```diff
@@ -10,14 +10,25 @@ typedef struct {
+  String title;           ///< Title for progress message
+  String status;          ///< Status for progress message
+  DictOf(String, Object) extra_info;  ///< Extra info for 'echo' messages
+} MessageExtData;
```
it's not related to "ext". it's just messages. also not sure why we are flip-flopping between "Message" and "Msg". Let's just stick with Msg, even though some old names use "Message".

```suggestion
} MsgData;
```

justinmk commented on src/nvim/message_defs.h:
```diff
@@ -10,14 +10,25 @@ typedef struct {
+} MessageExtData;
 /// Message history for `:messages`
 typedef struct msg_hist {
+  MsgID msg_id;           ///< Indentifier of the message
```
```suggestion
  MsgID msg_id;           ///< Message id.
```

justinmk commented on runtime/lua/vim/pack.lua:
```diff
@@ -395,7 +395,7 @@ local function new_progress_report(title)
     local details = (' %s %s'):format(title, fmt:format(...))
     local chunks = { { 'vim.pack', 'ModeMsg' }, { ': ' }, { progress, 'WarningMsg' }, { details } }
-    vim.api.nvim_echo(chunks, true, { kind = 'progress' })
+    vim.api.nvim_echo(chunks, true, {})
```
does it need a `-- TODO` comment ?

justinmk commented on src/nvim/api/keysets_defs.h:
```diff
@@ -336,6 +336,11 @@ typedef struct {
+  String title;
+  String status;
+  Integer percent;
+  DictOf(Object) extra_info;
```
```suggestion
  DictOf(Object) data;
```

justinmk commented on runtime/doc/api.txt:
```diff
@@ -678,6 +678,7 @@ nvim_echo({chunks}, {history}, {opts})                           *nvim_echo()*
                        by listening for the `Progress` event
                    • percent: How much progress is done on the progress
                      message
+                   • extra_info: dictionary containing additional information
```
> if we should allow something like this in all kinds of messages. Kind can be set to any arbitrary value. If we let arbitrary info be passed. ext-ui and plugins can add support for out-of-spec communication if they need to. ¯_(ツ)_/¯ .

yes, anything that applies to all messages should not be restricted to progress messages. (if that requires extra work, please create a tracking issue)

also it's somewhat of a stability risk: a plugin could (accidentally) store a gigantic object on the message? need a tracking issue for that too.

justinmk commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,210 @@ it('pager works in headless mode with UI attached', function()
+    setup_autocmd()
+  end)
+
+  it('can send progress-message through nvim-echo', function()
```
```suggestion
  it('can send progress-message through nvim_echo', function()
```

justinmk commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,210 @@ it('pager works in headless mode with UI attached', function()
+    )
+  end)
+
+  it('can send arbitrary data through extra_info', function()
```
```suggestion
  it('user-defined data in `data` field', function()
```

justinmk commented on src/nvim/message.c:
```diff
@@ -1032,16 +1078,53 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
 
-  // Allocate an entry and add the message at the end of the history.
-  MessageHistoryEntry *entry = xmalloc(sizeof(MessageHistoryEntry));
+  bool is_kind_progress = strequal(msg_ext_kind, MSG_KIND_PROGRESS);
```
```suggestion
  bool is_progress = strequal(msg_ext_kind, MSG_KIND_PROGRESS);
```

justinmk commented on src/nvim/message.c:
```diff
@@ -1012,18 +1020,56 @@ static void msg_hist_add(const char *s, int len, int hl_id)
+  return entry;
+}
+
+static void emit_progress_event(MessageHistoryEntry *msg)
```
this is the C codebase convention, AFAICT. (the `_event` suffix is for the event-loop, something totally different)
```suggestion
static void do_autocmd_progress(MessageHistoryEntry *msg)
```

justinmk commented on src/nvim/message.c:
```diff
@@ -1032,16 +1078,53 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
-  // Allocate an entry and add the message at the end of the history.
-  MessageHistoryEntry *entry = xmalloc(sizeof(MessageHistoryEntry));
+  bool is_kind_progress = strequal(msg_ext_kind, MSG_KIND_PROGRESS);
+  bool old_message_found = false;
```
```suggestion
  bool old_msg_found = false;
```

justinmk commented on runtime/doc/api-ui-events.txt:
```diff
@@ -913,4 +932,29 @@ must handle.
 	    True when sent with |g<| command, false with |:messages|.
 
+==============================================================================
+PROGRESS MESSAGE                                    *progress-message*
```
This help file is for the UI protocol. Progress messages are not specific to the UI protocol (though they happen to emit UI events, just like any other regular message).

Add this as section "4." after this: https://github.com/neovim/neovim/blob/58060c2340a52377a0e1d2b782ce1deef13b2b9b/runtime/doc/message.txt#L754-L755

justinmk commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,210 @@ it('pager works in headless mode with UI attached', function()
   ]])
 end)
+
+describe('progress-message', function()
```
add a `it('validates')` block that tests all/most of the errors https://github.com/neovim/neovim/blob/58060c2340a52377a0e1d2b782ce1deef13b2b9b/test/functional/lua/vim_spec.lua#L2921-L2925

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -787,6 +787,31 @@ ModeChanged			After changing the mode. The pattern is
+				`nvim_echo`. The pattern is matched against
+				title of the message. The |event-data| contains:
+					msg_id: id of the message
+					content: text of the message
```
```suggestion
					text: text of the message
```

shadmansaleh replied:
msg_show ui+event already uses content for this.  Should we give a different name for the autocmd event?

justinmk reviewed (commented):
After comments are resolved I think we can merge. Please focus only on resolving comments, and create tracking issues for any open questions, so that we can close this PR.

shadmansaleh commented on src/nvim/message.c:
```diff
@@ -1032,16 +1078,53 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
+    old_message_found = true;
+    // detach the node if found
+    if (entry->prev) {
+      entry->prev->next = entry->next;
+    } else {
+      msg_hist_first = entry->next;
+    }
+    if (entry->next) {
+      entry->next->prev = entry->prev;
+    } else {
+      msg_hist_last = entry->prev;
```
This was actually a question from the begining. Should we put the msg at end of history once it get's updated or leave it where it is and change the message? I'm kind of more inclined toward not moving the msg in history sequence. Which one would be better?

shadmansaleh commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,210 @@ it('pager works in headless mode with UI attached', function()
+    screen:expect([[
+      ^                                        |
+      {1:~                                       }|*3
+      test-message                            |
```
The events still get the percent updates. And they are being asserted in the test. It's just not yet shown in tui yet. Will need a followup to update to ext_ui2

shadmansaleh commented on src/nvim/message.c:
```diff
@@ -149,6 +150,8 @@ bool keep_msg_more = false;    // keep_msg was set by msgmore()
 
 // Extended msg state, currently used for external UIs with ext_messages
 static const char *msg_ext_kind = NULL;
+static MsgID msg_ext_id = 0;
+static DictOf(Object) msg_ext_progress = ARRAY_DICT_INIT;
```
Other variables for msg_show are also named with ext. I thought it meant ext-ui. Should we not follow that scheme?

shadmansaleh commented on src/nvim/message.c:
```diff
@@ -158,6 +161,8 @@ static bool msg_ext_history = false;  ///< message was added to history
 
 static int msg_grid_pos_at_flush = 0;
 
+static MsgID msg_id_next = 1;           ///< message id to be allocated to next message
```
msg_ext_id is basically msg-id of current message being shown (to be sent to ui) . the ext_ variables basically act as buckets for data to be sent to through msg_show. Echo places data in the variables then on other side when msg-show event is released it pickes data from here. They kind of holds the data and other piece of code flushes it once event is released.

Msg_id_next is next unique-id for message. It get's increamented everytime a message in created and assigned an id. Because when msg gets updated id is reused msg-id-next and msg-ext-id aren't same

justinmk replied:
> msg_ext_id is basically msg-id of current message being shown (to be sent to ui)

all messages should have IDs. that is not related to UI protocol nor `ext_messages`. it's just part of messages.

the use of `ext_` in some of these variable names I think is not totally coherent. for the message-id concept it definitely is unrelated.



> to be sent to through msg_show

msg_show happens for all messages, by definition. it does not mean that everything message-related is somehow part of `ext_messages` / UI protocol.

shadmansaleh replied:
I think there has been a bit of confusion regarding this.

msg_ext_id is for msg_show ui-event and is part of the following group of variables
https://github.com/neovim/neovim/blob/516363e6baa0329dac4b048526e3deadb0d580d0/src/nvim/message.c#L151-L155

msg_id_next is to track which next message-id we can give.
They are different things.

The `ext` on the structs and data was wrong. I was using ext as short for extra there. That was wrong. I've renamed those.

shadmansaleh committed:
address-review 2

shadmansaleh committed:
bugfix + add test addressing review

shadmansaleh committed:
fix lint

shadmansaleh committed:
fix lintc

shadmansaleh committed:
document data field for ui_show event

shadmansaleh commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3416,4 +3416,38 @@ describe('progress-message', function()
     )
   end)
 
+  it('gets moved to end of history upon update', function()
```
As far as I remember, the discussion was, progress events are really close to messages. So keeping them together keeps us from having separate structures/implementations for two similar concepts.

An ID was necessary to identify which message to update.

things like title, status, percent ar useful because they gives External UIs flexibility to make interesting UIs. For example, a progress bar to show the percent done instead of text. putting that info in msg makes something like that hard.

Although we dropped partial updates for now. We could probably get away without storing them.

shadmansaleh committed:
don't store title, status, percent, data fields in msg_hist

shadmansaleh committed:
perform null check before dereferencing msg_data

justinmk commented:
Still a [naming nit](https://github.com/neovim/neovim/pull/34846#discussion_r2298874846) remaining, then I think this is ready? @luukvbaal does it LGTY ?

luukvbaal commented on runtime/doc/api-ui-events.txt:
```diff
@@ -881,6 +882,24 @@ must handle.
 
+	msg_id
+	    Unique identifier for the message in history. Is greater or equal
+	    to 1. 0 indicates msg_id was not assigned. When message of same id
```
Always assigning/emitting an ID would be less surprising. Though I suppose this `0` conveys that no caller is aware of the msg->id mapping (and thus shouldn't be replaced in the future) unlike `nvim_echo` emitted messages? Though given https://github.com/neovim/neovim/pull/34846/files#r2299325332 we might be better off just always assigning an ID as suggested in https://github.com/neovim/neovim/issues/32537#issuecomment-2672600804:

> All messages (incl. non-progress) have an id."

Also `msg_` is redundant here. Better use `id`?

shadmansaleh replied:
autocmd events themselve has a id feid. ti becomes something like

```
{
id: ,
...
data: {
   id: ,
   ....
}
}
```

I thought that was confusing so the `msg_`

luukvbaal commented on runtime/doc/api.txt:
```diff
@@ -659,13 +659,32 @@ nvim_echo({chunks}, {history}, {opts})                           *nvim_echo()*
+        (`integer`) Message id. Valid message id is always greater then 0
+        • -1 means nvim_echo didn't show a message
+        • 0 means nvim_echo didn't allocate a message id for the message.
+          happens for temp messages not stored in message history.
```
Linking the message ID to whether or not a message is stored in the history seems wrong to me. There is no reason a temporary message shouldn't be able to be replaced in the UI.

shadmansaleh replied:
If we give IDs to temp messages, then we can't really validate the ids. We'll just have to trust the user and forward it.

luukvbaal reviewed (commented):
We should determine whether changing the :messages history for updated messages is actually wanted. I would say it is not. All that achieves is obfuscating the actual message history?

If we leave the history intact a UI could prepend the message ID when displaying the message history to actually convey that a message was updated.

(That would also mean the ID is not used internally and only emitted as part of the msg_show/Progress events, which to me would be a welcome simplification.)

shadmansaleh committed:
keep all copies of progress message on history

shadmansaleh committed:
rename content -> text in autocmd docs

shadmansaleh commented on runtime/doc/api-ui-events.txt:
```diff
@@ -881,6 +882,24 @@ must handle.
+	            • cancel: The progressing process should be canceled.
+	       • percent: How much progress is done on the progress
+                     message
+	       • data: dictionary containing additional information.
```
UIs can still give filtering options based on data. Or provide some way to say where to put them. like adding a location arg. I think we should keep it. it's already taken from the user. We are just forwarding it.

shadmansaleh committed:
allocate msg-ids for messages not in history too

shadmansaleh committed:
add support for string ids

shadmansaleh committed:
remove progress.data field from msg_show event

shadmansaleh commented:
> unreleased APIs can change on master (and tests won't fail)

By unreleased do you mean unreleased changes to api?

Well either way it works now with string ids too :]

justinmk commented:
> (That would also mean the ID is not used internally and only emitted as part of the msg_show/Progress events, which to me would be a welcome simplification.)

I would like to keep the door open for retrieving messages by id from the message history, even if we don't expose that now. Do you foresee a problem with that?

justinmk commented on runtime/doc/autocmd.txt:
```diff
@@ -787,6 +787,31 @@ ModeChanged			After changing the mode. The pattern is
 		:au ModeChanged *:[vV\x16]* let &l:rnu = mode() =~# '^[vV\x16]'
 		:au WinEnter,WinLeave * let &l:rnu = mode() =~# '^[vV\x16]'
+Progress						*Progress*
+				After a progress message is created via
```
```suggestion
				When a progress message was created or updated via
```

shadmansaleh committed:
address review (update autocmd-doc)

justinmk commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,385 @@ it('pager works in headless mode with UI attached', function()
+  it('validates', function()
+    -- throws error if title, status, percent, data is used in non progress message
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
```
```suggestion
      'title, status, percent and data fields can only be used with progress messages',
```

justinmk commented on test/functional/ui/messages_spec.lua:
```diff
@@ -3138,3 +3138,385 @@ it('pager works in headless mode with UI attached', function()
+
+    eq(
+      'title, status, percents and data fields can only be used with progress messages',
+      t.pcall_err(api.nvim_echo, { { 'test-message' } }, false, { data = { tag = 'test' } })
```
why are we restricting `data` to Progress messages? I guess because we don't have a `Message` event yet.

shadmansaleh replied:
Yeah,, for now, data doesn't go anywhere except for progress-event. if we add a general message event we can open it up for other kinds.

justinmk commented on src/nvim/message.c:
```diff
@@ -1061,6 +1109,24 @@ static void msg_hist_add_multihl(HlMessage msg, bool temp)
+  }
+
+  if (is_progress) {
+    do_autocmd_progress(msg_id, entry, msg_data);
```
not a blocker for this PR but hopefully can resolve it as a followup :

executing user code in random parts of the C core is risky and always leads to bugs.
is it possible to move this somewhere closer to where user code is *already* executed, later in this codepath perhaps? 

ideally by implementing https://github.com/neovim/neovim/issues/35366

shadmansaleh replied:
I'm not sure what you mean by later in codepath. It's already almost at the end. After this, we don't do much other then cleanup and function return.

A safe environment for executing autocommands would be nice. I believe it's out of scope for this pr.

justinmk replied:
> After this, we don't do much other then cleanup and function return.

somewhere after the function return. i think the most sane approach is https://github.com/neovim/neovim/issues/35366 , but yes it's out of scope for now, though i don't think it will be hard.

luukvbaal replied:
56cbe8c works but is not what justin was getting at. #35366 talks about queuing events and finding a safe state in the event loop to execute them (outside of the function calls that queue them).

shadmansaleh replied:
Yeah, that issue is a different thing. This pr is not trying to solve that issue. I was asking about the other mitigation justinmk was talking about : ]

shadmansaleh committed:
address review 4 (fix typo in error message for percent)

justinmk reviewed (approved):
LGTM,  @luukvbaal ?

shadmansaleh committed:
emit progress event from nvim_echo function

luukvbaal reviewed (approved):
>  I would like to keep the door open for retrieving messages by id from the message store, even if we don't expose that now. Do you foresee a problem with that?

No that seems fine. Actually to forward the IDs to the `msg_history_show` event we would have to store them😅 IDK if UIs will use those, but until now we have kept the `msg_(history_)show` events in sync with new relevant arguments. At least the ID seems relevant, but maybe the progress field is as well (sorry about that)...

shadmansaleh committed:
remove unnecesssary MSG_KIND_PROGRESS macro

shadmansaleh commented:
Did our CI die :| ?

clason commented on runtime/lua/vim/pack.lua:
```diff
@@ -395,7 +395,8 @@ local function new_progress_report(title)
     local chunks = { { 'vim.pack', 'ModeMsg' }, { ': ' }, { progress, 'WarningMsg' }, { details } }
-    vim.api.nvim_echo(chunks, true, { kind = 'progress' })
+    -- TODO: need to add support for progress-messages api
+    vim.api.nvim_echo(chunks, true, {})
```
Yes, and I am removing those currently. So I'm pointing out an obvious regression when you touch the code.

shadmansaleh committed:
use api instead of vim.api when available

justinmk commented on src/nvim/api/vim.c:
```diff
@@ -758,14 +758,31 @@ void nvim_set_vvar(String name, Object value, Error *err)
+///          - data: dictionary containing additional information
+/// @return Message id.
+///         - -1 means nvim_echo didn't show a message
+Union(Integer, String) nvim_echo(ArrayOf(Tuple(String, *HLGroupID)) chunks, Boolean history,
```
Thanks, yes, I found a test for the string id case. Should we disallow arbitrary integers though? That seems weird. If a caller specifies an arbitrary integer, then another caller which allows the id to be generated, could accidentally use it?
